File: cpp/0007.reverse-integer.cpp
class Solution
{
public:
    int reverse(int x)
    {
        if (!x)
        {
            return x;
        }
        bool isNegative = (x < 0);
        if (x < 0)
        {
            x *= -1;
        }
        int res32 = 0;
        long long res64 = 0;
        for (; x; x /= 10)
        {
            int v = x % 10;
            res32 = res32 * 10 + v;
            res64 = res64 * 10 + v;
        }
        if (isNegative)
        {
            res32 *= -1;
            res64 *= -1;
        }
        if ((long long)res32 != res64)
        {
            return 0;
        }
        return res32;
    }
};

File: rust/0007_reverse-integer.rs
impl Solution
{
    pub fn reverse(x: i32) -> i32
    {
        let mut is_negative = false;
        let mut y32 = 0;
        let mut y64 = 0 as i64;
        let mut x32 = x;

        if x32 < 0
        {
            is_negative = true;
            x32 *= -1;
        }

        while x32 > 0
        {
            let n = x32 % 10;

            y32 = y32 * 10 + n;
            y64 = y64 * 10 + n as i64;
            x32 /= 10;
        }

        if is_negative
        {
            y32 *= -1;
            y64 *= -1;
        }

        if y32 as i64 != y64
        {
            y32 = 0;
        }

        return y32;
    }
}
